compiler/noirc_evaluator/src/ssa/interpreter/intrinsics.rs

```diff
@@ -776,37 +774,56 @@ fn new_embedded_curve_point(
     ))
 }
 
-/// Convert a [Value] to a vector of [FieldElement] for printing.
-fn value_to_fields(value: &Value) -> Vec<FieldElement> {
-    fn go(value: &Value, fields: &mut Vec<FieldElement>) {
-        match value {
-            Value::Numeric(numeric_value) => fields.push(numeric_value.convert_to_field()),
-            Value::Reference(reference_value) => {
-                if let Some(value) = reference_value.element.borrow().as_ref() {
-                    go(value, fields);
+/// Convert a slice of [Value] to a flattened vector of [FieldElement] for printing.
+///
+/// It takes a slice, rather than individual values, so that it can try to
+/// pair up `u32` fields indicating the size of a `Slice` with its elements
+/// following in the next value, and truncate the data to the semantic length.
+fn values_to_fields(values: &[Value]) -> Vec<FieldElement> {
+    fn go<'a, I>(values: I, fields: &mut Vec<FieldElement>)
+    where
+        I: Iterator<Item = &'a Value>,
+    {
+        let mut vector_length: Option<usize> = None;
+        for value in values {
+            match value {
+                Value::Numeric(numeric_value) => fields.push(numeric_value.convert_to_field()),
+                Value::Reference(reference_value) => {
+                    if let Some(value) = reference_value.element.borrow().as_ref() {
+                        go(std::iter::once(value), fields);
+                    }
                 }
-            }
-            Value::ArrayOrSlice(array_value) => {
-                for value in array_value.elements.borrow().iter() {
-                    go(value, fields);
+                Value::ArrayOrSlice(array_value) => {
+                    let length = match vector_length {
+                        Some(length) if array_value.is_slice => {
+                            length * array_value.element_types.len()
+                        }
+                        _ => array_value.elements.borrow().len(),
+                    };
+                    go(array_value.elements.borrow().iter().take(length), fields);
+                }
+                Value::Function(id) => {
+                    // Based on `decode_printable_value` it will expect consume the environment as well,
+                    // but that's catered for the by the SSA generation: the env is passed as separate values.
+                    fields.push(FieldElement::from(id.to_u32()));
```

Comment:
I think `vector_length` must also be set to `None` here.

Reply:
It happens for all cases on line 821: we either have a number, and then set the length to that, or we don't, because we just used it or had something else, either way it's being reset.